power_grid_model_c/power_grid_model/include/power_grid_model/math_solver/y_bus.hpp

```diff
@@ -415,18 +497,65 @@ template <bool sym> class YBus {
         return shunt_flow;
     }
 
+    /// @brief register a new callback to signal a parameter change
+    /// @param callback the callback to register
+    /// @return the unique key referencing this callback (used for unregistering)
+    uint64_t register_parameters_changed_callback(ParamChangedCallback callback) {
+        static uint64_t num_added = 0;
+
+        auto const new_key = num_added++;
+        assert(!parameters_changed_callbacks_.contains(new_key));
+        parameters_changed_callbacks_.insert(std::make_pair(new_key, std::move(callback)));
+        return new_key;
+    }
+
+    /// @brief unregister a callback to signal a parameter change
+    /// @param key the unique key referencing the callback (returned by register_parameters_changed_callback)
+    void unregister_parameters_changed_callback(uint64_t key) {
+        assert(parameters_changed_callbacks_.contains(key));
+        parameters_changed_callbacks_.erase(key);
+    }
+
   private:
     // csr structure
     std::shared_ptr<YBusStructure const> y_bus_struct_;
 
     // admittance
-    std::shared_ptr<ComplexTensorVector<sym> const> admittance_;
+    std::shared_ptr<ComplexTensorVector<sym>> admittance_;
```

Comment:
Can you clarify the design choices here? We now allow multiple threads to modify the admittance at the same time? How can we prevent data racing?

Reply:
As I understand it, with each instance of the model being copied, the `y_bus` is checked to see if it is necessary to make a copy. Other cases where the `admittance_` matrix is involved, it is accessed const.

Reply:
yes, sure. The choice was the following:

* we need to be able to actually change the ybus values in place (otherwise, **all** the matrix elements need to be regenerated every single parameter change)
* to do so, we need this to be not `const`
* data races are prevented by the check in the incremental update that `use_count()` (= the amount of references to the shared pointer) is identically 1
* if that check does not hold, a copy of the existing admittance matrix elements is made. this ensures no data races.
* in the next iteration, the `use_count()` is `1` again (due to the copy) and we may modify in-place

Reply:
```suggestion
    ComplexTensorVector<sym> admittance_;
```

Reply:
fixed